app/models/manageiq/providers/amazon/inventory/parser/cloud_manager.rb

```diff
@@ -317,29 +317,51 @@ def hardware_disks(persister_hardware, instance, flavor)
-        :name                     => flavor[:name],
-        :description              => flavor[:description],
-        :enabled                  => !flavor[:disabled],
-        :cpus                     => flavor[:vcpu],
-        :cpu_cores                => 1,
-        :memory                   => flavor[:memory],
-        :supports_32_bit          => flavor[:architecture].include?(:i386),
-        :supports_64_bit          => flavor[:architecture].include?(:x86_64),
-        :supports_hvm             => flavor[:virtualization_type].include?(:hvm),
-        :supports_paravirtual     => flavor[:virtualization_type].include?(:paravirtual),
-        :block_storage_based_only => flavor[:ebs_only],
-        :cloud_subnet_required    => flavor[:vpc_only],
-        :ephemeral_disk_size      => flavor[:instance_store_size],
-        :ephemeral_disk_count     => flavor[:instance_store_volumes]
+      persister.flavors.find_or_build(flavor['instance_type']).assign_attributes(
+        :name                     => flavor['instance_type'],
+        :description              => flavor['instance_type'],
+        :enabled                  => true,
+        :cpus                     => flavor.dig('v_cpu_info', 'default_v_cpus'),
+        :cpu_cores                => flavor.dig('v_cpu_info', 'default_cores') || 1,
+        :memory                   => flavor.dig('memory_info', 'size_in_mi_b')&.megabytes,
+        :supports_32_bit          => flavor.dig('processor_info', 'supported_architectures').include?('i386'),
+        :supports_64_bit          => flavor.dig('processor_info', 'supported_architectures').include?('x86_64'),
+        :supports_hvm             => true,
+        :supports_paravirtual     => supports_paravirtual?(flavor),
+        :block_storage_based_only => flavor.dig('supported_root_device_types').any? { |device| device != 'ebs' },
+        :ephemeral_disk_size      => flavor.dig('instance_storage_info', 'total_size_in_gb')&.gigabytes || 0,
+        :ephemeral_disk_count     => flavor.dig('instance_storage_info', 'disks')&.size || 0,
+        :publicly_available       => true,
+        :cloud_subnet_required    => vpc_only?(flavor)
       )
     end
   end
 
+  # From the Amazon online documentation: "The following previous generation instance types
+  # support PV AMIs: C1, C3, HS1, M1, M3, M2, and T1. Current generation instance types
+  # do not support PV AMIs."
+  #
+  def supports_paravirtual?(flavor)
+    return false if flavor['current_generation']
+    return false unless %w[c1 c3 hs1 m1 m3 m2 t1].include?(flavor['instance_type'].split('.').first.downcase)
+
+    true
+  end
+
+  # This logic is cobbled together from various online sources. The V3 API does not include
+  # such information, but most current flavors are vpc-only from what I can gather.
+  #
+  def vpc_only?(flavor)
+    if %w[a1 c1 c3 r3].include?(flavor['instance_type'].split('.').first.downcase)
```

Comment:
Minor, but I think this `.split('.').first.downcase` bit should be moved into a private method to make it clearer what it's doing (and also because it's used up above in the supports_paravirtual method as well).  I'm thinking something like:

```ruby
private def instance_family(flavor)
  flavor.split('.').first.downcase
end
```

Reply:
Done.